src/graphics/vulkan.zig

```diff
@@ -0,0 +1,150 @@
+	VK_THREAD_DONE_KHR = 1000268001,
+	VK_OPERATION_DEFERRED_KHR = 1000268002,
+	VK_OPERATION_NOT_DEFERRED_KHR = 1000268003,
+	VK_ERROR_INVALID_VIDEO_STD_PARAMETERS_KHR = -1000299000,
+	VK_ERROR_COMPRESSION_EXHAUSTED_EXT = -1000338000,
+	VK_PIPELINE_BINARY_MISSING_KHR = 1000483000,
+	VK_ERROR_NOT_ENOUGH_SPACE_KHR = -1000483000,
+	VK_ERROR_INCOMPATIBLE_SHADER_BINARY_EXT = 1000482000,
+	VK_RESULT_MAX_ENUM = 2147483647,
+};
+
+pub fn checkResult(result: c.VkResult) void {
+	const resultEnum = std.meta.intToEnum(VkResultEnum, result) catch {
+		std.log.err("Encountered a vulkan error with unknown error code {}", .{result});
+		return;
+	};
+	if(resultEnum == .VK_SUCCESS) return;
+	std.log.err("Encountered a vulkan error: {s}", .{@tagName(resultEnum)});
+}
+
+fn checkResultIfAvailable(result: anytype) void {
+	if(@TypeOf(result) != void) {
+		checkResult(result);
+	}
+}
+
+fn allocEnumerationGeneric(function: anytype, allocator: NeverFailingAllocator, args: anytype) []@typeInfo(@typeInfo(@TypeOf(function)).@"fn".params[@typeInfo(@TypeOf(function)).@"fn".params.len - 1].type.?).pointer.child {
+	const T = @typeInfo(@typeInfo(@TypeOf(function)).@"fn".params[@typeInfo(@TypeOf(function)).@"fn".params.len - 1].type.?).pointer.child;
+	var count: u32 = 0;
+	checkResultIfAvailable(@call(.auto, function, args ++ .{&count, null}));
+	const list = allocator.alloc(T, count);
+	checkResultIfAvailable(@call(.auto, function, args ++ .{&count, list.ptr}));
+	return list;
+}
+
+pub fn enumerateInstanceLayerProperties(allocator: NeverFailingAllocator) []c.VkLayerProperties {
+	return allocEnumerationGeneric(c.vkEnumerateInstanceLayerProperties, allocator, .{});
+}
+
+pub fn enumerateInstanceExtensionProperties(allocator: NeverFailingAllocator, layerName: ?[*:0]u8) []c.VkExtensionProperties {
+	return allocEnumerationGeneric(c.vkEnumerateInstanceExtensionProperties, allocator, .{layerName});
+}
+
+pub const Instance = struct { // MARK: Instance
+	var instance: c.VkInstance = undefined;
+
+	const validationLayers: []const [*:0]const u8 = &.{
+		"VK_LAYER_KHRONOS_validation",
+	};
+
+	fn checkValidationLayerSupport() bool {
+		const availableLayers = enumerateInstanceLayerProperties(main.stackAllocator);
+		defer main.stackAllocator.free(availableLayers);
+		for(validationLayers) |layerName| continueOuter: {
+			for(availableLayers) |layerProperties| {
+				if(std.mem.eql(u8, std.mem.span(layerName), std.mem.span(@as([*:0]const u8, @ptrCast(&layerProperties.layerName))))) {
+					break :continueOuter;
+				}
+			}
+			std.log.warn("Couldn't find validation layer {s}", .{layerName});
+			return false;
+		}
+		return true;
+	}
+
+	pub fn init() void {
+		if(c.gladLoaderLoadVulkan(null, null, null) == 0) {
+			@panic("GLAD failed to load Vulkan functions");
+		}
+		const appInfo = c.VkApplicationInfo{
+			.sType = c.VK_STRUCTURE_TYPE_APPLICATION_INFO,
+			.pApplicationName = "Cubyz",
+			.applicationVersion = c.VK_MAKE_VERSION(0, 0, 0),
+			.pEngineName = "custom",
```

Comment:
Since this field is used to identify game and engine when applying optimizations from driver side I would suggest `Cubyz Engine` instead of generic `custom`. Although that does not matter, looking after looking on [Mesa](https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/util/00-radv-defaults.conf) code.

Reply:
Or we can set it to `null` since this field is nullable and we don't really have an named engine.